optuna/samplers/_qmc.py

```diff
@@ -0,0 +1,350 @@
+from collections import OrderedDict
+from typing import Any
+from typing import Dict
+from typing import Optional
+
+import numpy
+import scipy
+
+import optuna
+from optuna import distributions
+from optuna import logging
+from optuna._experimental import experimental
+from optuna._transform import _SearchSpaceTransform
+from optuna.distributions import BaseDistribution
+from optuna.samplers import BaseSampler
+from optuna.study import Study
+from optuna.trial import FrozenTrial
+from optuna.trial import TrialState
+
+
+_logger = logging.get_logger(__name__)
+
+_SUGGESTED_STATES = (TrialState.COMPLETE, TrialState.PRUNED)
+
+_NUMERICAL_DISTRIBUTIONS = (
+    distributions.UniformDistribution,
+    distributions.LogUniformDistribution,
+    distributions.DiscreteUniformDistribution,
+    distributions.IntUniformDistribution,
+    distributions.IntLogUniformDistribution,
+)
+
+
+@experimental("2.x.0")  # TODO(kstoneriv3)
+class QMCSampler(BaseSampler):
+    """A Quasi Monte Carlo Sampler that generates low-discrepancy sequences.
+
+    Quasi Monte Carlo (QMC) sequences are designed to have low-discrepancies than
+    standard random seqeunces. They are known to perform better than the standard
+    randam sequences in hyperparameter optimization.
+
+    For further information about the use of QMC sequences for hyperparameter optimization,
+    please refer to the following paper:
+
+    - Bergstra, James, and Yoshua Bengio. Random search for hyper-parameter optimization.
+      Journal of machine learning research 13.2, 2012.
+      <https://jmlr.org/papers/v13/bergstra12a.html>`_
+
+    We use the QMC implementations in Scipy. For the details of the QMC algorithm,
+    see the Scipy API references on `scipy.stats.qmc
+    <https://scipy.github.io/devdocs/stats.qmc.html>`_.
+    TODO(kstoneriv3): make some notes on 2^k property of Sobol and prime property of Halton
+
+    .. note:
+        Please note that this sampler does not support CategoricalDistribution.
+        If your search space contains categorical parameters, it samples the catagorical
+        parameters by its `independent_sampler` without using QMC algorithm.
+
+    Args:
+        qmc_type:
+            The type of QMC sequence to be sampled. This must be one of
+            `"sobol"`, `"halton"`, `"LHS"` and `"OA-LHS"`. Default is `"sobol"`.
+
+        scramble:
```

Comment:
Should you actually allow this? There is not really a reason to use the unscrambled version appart from making academic papers about the bare sequence IMO. The convergence properties are better using scrambling.

Reply:
Hmm, I just tried to add a wrapper to Scipy so maybe this is unnecessary for hyperparameter optimization...

The only problem for not supporting `scramble` would be that parallel / distributed workers will no longer be synchronized by the default setting. Currently, all the workers have `scramble=False` by default so their samples are from the same QMC sequence but if we get rid of `scramble`, they will have different seeds by default and the samples will not be from the same sequence.

Reply:
If the `seed` that you give is not shared, there would be no issue as `scipy.stats.qmc` handles it properly.

Reply:
> If the `seed` that you give is not shared, there would be no issue as `scipy.stats.qmc` handles it properly.

I mostly agree with removing the `scramble` option, but I do not understand what you mean by this...

Reply:
As I got your code, you reuse the same QMC engine no? Because you check the number of existing sample, fastforward and ask for the next value. So I don't really understand why you would have an issue actually.

If you have different samplers, then you just need to initialise everyone with the same original seed (deepcopy can do the trick). Then with the same seed you are ensure to have the exact same sequence for everyone.

Reply:
> As I got your code, you reuse the same QMC engine no? Because you check the number of existing sample, fastforward and ask for the next value. So I don't really understand why you would have an issue actually.

Your understanding seems correct, but I do not get what you mean by "an issue"...

> If you have different samplers, then you just need to initialise everyone with the same original seed (deepcopy can do the trick). Then with the same seed you are ensure to have the exact same sequence for everyone.

The problem is that if the user might forget to provide the same seed in such distributed/parallel cases, and in such cases the samplers will have different seeds as the default seed is `seed=None`. I believe it is more user-friendly to let samplers behave without scramble to avoid such misuse.

Reply:
As so you mean that the whole thing will get called with `seed=None` multiple times? Ok I didn't understood this...
Then I guess there is not really an option if you cannot globally enforce a common seed even if the user does not provide one.

Reply:
Optuna supports distributed optimization with multiple machines (in such cases, information of optimization is stored in the RDB server). Then, the sampler in each machine will be initialized independently so they will have different seeds.